### PR TITLE
Makes history navigation stop at top and bottom and and fixes several…

### LINKED ISSUE #4669: History in interactive window should not cycle
It should stop when it reaches 1st item in the history list and user tries to go to previous and similarly when it is at the last item and user tries to go to next. This is the behavior in command window too.

### PR DESCRIPTION
**Pull request by ManishJayaswal**
… history navigation bugs. Fixes #4669

### CODE REVIEW

**Comment by ManishJayaswal** [comment]
FYI @amcasey @cston @KevinH-MS @tmat

File: src/InteractiveWindow/Editor/History.cs [hunk 1]
```diff
@@ -159,11 +164,16 @@ private Entry Get(string pattern, int step)
 
         internal Entry GetNext(string pattern = null)
         {
+            //if you are at the last history entry or history navigation pointer is
+            //not initialized then return null as there is no next entry
+            if ((_current == Length - 1) || (_current == -1)) return null;
             return Get(pattern, step: +1);
         }
 
         internal Entry GetPrevious(string pattern = null)
```
Comment (cston) [inline src/InteractiveWindow/Editor/History.cs]: Not part of this change but the optional parameter value is not necessary in GetPrevious or GetNext.

File: src/InteractiveWindow/Editor/History.cs [hunk 2]
```diff
@@ -159,11 +164,16 @@ private Entry Get(string pattern, int step)
...
         {
+            //if at first entry then return null
+            if (_current == 0) return null;
             return Get(pattern, step: -1);
```
Comment (cston) [inline src/InteractiveWindow/Editor/History.cs]: Is it possible for _current == -1 here and will Get(pattern, -1) handle that? (Hit previous, change the text and submit, hit previous.)

**Comment by cston** [comment]
LGTM

File: src/InteractiveWindow/Editor/History.cs [hunk 3]
```diff
@@ -159,11 +164,16 @@ private Entry Get(string pattern, int step)
 
         internal Entry GetNext(string pattern = null)
         {
+            //if you are at the last history entry or history navigation pointer is
+            //not initialized then return null as there is no next entry
+            if ((_current == Length - 1) || (_current == -1)) return null;
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/History.cs]: Would it make more sense to remove the wraparound logic from ```Move``` (new line ~219)?

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs [hunk 4]
```diff
@@ -483,6 +495,245 @@ public void CheckHistoryPrevious()
         }
 
         [Fact]
+        public void CheckHistoryPreviousNotCircular()
+        {
+            //submit, submit, up, up, up
+            const string inputString1 = "1 ";
+            const string inputString2 = "2 ";
+            Window.InsertCode(inputString1);
```
Comment (amcasey) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: Is the goal is to create history entries, would it make more sense to call ```AddInput```?

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs [hunk 5]
```diff
@@ -483,6 +495,245 @@ public void CheckHistoryPrevious()
...
+            Assert.Equal(inputString1, GetTextFromCurrentLanguageBuffer());
+            Task.Run(() => Window.Operations.ExecuteInput()).PumpingWait();
```
Comment (amcasey) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: Extract into a method?  We seem to do this a lot.

File: src/InteractiveWindow/Editor/History.cs (hunk 2 above)
Comment (ManishJayaswal) [inline src/InteractiveWindow/Editor/History.cs]: @cston  yes it handles it. The test "CheckHistoryPreviousAfterSubmittingNewEntryWhileNavigatingHistory" tests this scenario.

File: src/InteractiveWindow/Editor/History.cs (hunk 1 above)
Comment (ManishJayaswal) [inline src/InteractiveWindow/Editor/History.cs]: @cston you are right. I will change that.

File: src/InteractiveWindow/Editor/History.cs (hunk 3 above)
Comment (ManishJayaswal) [inline src/InteractiveWindow/Editor/History.cs]: @amcasey I thought move method was written with a goal to search entire history so I did not change it. We are certainly not using it to do so ATM but we may want to in future.

Comment (amcasey) [inline src/InteractiveWindow/Editor/History.cs]: Is ```Move``` really general purpose?  It has logic for stopping after a fixed number of steps which seems tied to a specific history browsing design.  Since it's a private method, I think it makes sense to remove the functionality we're not using and rebuild it later if our needs change.

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs (hunk 4 above)
Comment (ManishJayaswal) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: @amcasey AddInput does a bunch of other processing like making the buffer as readonly. So if the command is not actually submitted then the next call to AddInput fails. if we actually submit the command then it gets added to history twice - once in AddInput and once as part of executing the command. AddInput is doing much more than it's name suggests - we should perhaps rename it or refactor it.

File: src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs (hunk 5 above)
Comment (ManishJayaswal) [inline src/InteractiveWindow/EditorTest/InteractiveWindowTests.cs]: @amcasey done

File: src/InteractiveWindow/Editor/History.cs (hunk 3 above)
Comment (ManishJayaswal) [inline src/InteractiveWindow/Editor/History.cs]: @amcasey I was leaning towards doing it anyways and you just convinced me to do it. This is much cleaner. I moved all logic inside Move method.

File: src/InteractiveWindow/Editor/History.cs [hunk 6]
```diff
@@ -171,66 +176,84 @@ private Entry Move(string pattern, int step)
...
+                    if (TryMatch(pattern, out entry)) return entry;
                 }
+            }
+        }
 
-                visited++;
+        private bool TryMatch(string pattern, out Entry entry)
+        {
+            bool patternEmpty = string.IsNullOrWhiteSpace(pattern);
+            var _entry = _history[_current];
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/History.cs]: Why does this local name start with an underscore?

File: src/InteractiveWindow/Editor/History.cs [hunk 7]
```diff
@@ -171,66 +176,84 @@ private Entry Move(string pattern, int step)
...
-            int visited = 0;
-            while (true)
-            {
-                if (_current == end)
+                if (wasUninitialized) return null;
+             
+                while (true)
                 {
-                    if (visited < Length)
+                    if (_current == end)
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/History.cs]: Should this be ```>=```?  What happens if step is not 1?

Comment (amcasey) [inline src/InteractiveWindow/Editor/History.cs]: (Since it will be 1 in practice, feel free to assert that we actually saw equality.)

**Comment by ManishJayaswal** [comment]
@amcasey  addressed the CR feedback. Please take a look.

File: src/InteractiveWindow/Editor/History.cs [hunk 8]
```diff
@@ -155,82 +159,87 @@ private Entry Get(string pattern, int step)
...
-                start = Length - 1;
-                end = 0;
+                //return the current entry again ( handles case up, up, enter, up)
+                Entry entry;
+                if (TryMatch(pattern, out entry)) return entry;
             }
 
-            // no search in progress:
-            if (_current == -1)
+            for (_current--; _current >= end; _current--)
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/History.cs]: Did you mean to leave the decl position empty?
Also, I'd be tempted to eliminate ```end``` and just use ```0```.

File: src/InteractiveWindow/Editor/History.cs [hunk 9]
```diff
@@ -155,82 +159,87 @@ private Entry Get(string pattern, int step)
...
+                if (TryMatch(pattern, out entry)) return entry;
             }
 
-            bool patternEmpty = string.IsNullOrWhiteSpace(pattern);
+            return null;
+        }
 
-            int start, end;
-            if (step > 0)
+        private Entry MovePrevious(string pattern)
```
Comment (amcasey) [inline src/InteractiveWindow/Editor/History.cs]: I think you could inline ```MoveNext``` and ```MovePrevious``` but I'll understand if you prefer to leave that for later.

**Comment by amcasey** [comment]
:+1:

**Comment by ManishJayaswal** [comment]
Closing this as PR #4911 was used to make this change.
